spec/models/product_spec.rb

```diff
@@ -0,0 +1,52 @@
+require "rails_helper"
+
+RSpec.describe Product, type: :model do
+  describe "Associations" do
+    it{is_expected.to belong_to(:category)}
+    it{is_expected.to have_many(:order_details).dependent(:destroy)}
+    it{is_expected.to have_many(:orders).through(:order_details)}
+  end
+
+  describe "Validations" do
+    before {FactoryBot.build(:product)}
+    context "name" do
+      it{is_expected.to validate_presence_of(:name)}
+      it{is_expected.to validate_length_of(:name).is_at_most(Settings.str_length)}
+    end
+
+    context "description" do
+      it{is_expected.to validate_presence_of(:description)}
+    end
+
+    context "quantity" do
+      it{is_expected.to validate_presence_of(:quantity)}
+      it{is_expected.to validate_numericality_of(:quantity).only_integer.is_greater_than(Settings.zero)}
+    end
+
+    context "price" do
+      it{is_expected.to validate_presence_of(:price)}
+      it{is_expected.to validate_numericality_of(:quantity).is_greater_than(Settings.zero)}
+    end
+
+    context "unit" do
+      it{is_expected.to validate_presence_of(:unit)}
+    end
+  end
+
+  describe "scope" do
+    let!(:product){FactoryBot.create :product}
```

Comment:
Phân biệt let và let!

Reply:
khai báo let thì giá trị sẽ được cache lại trong một khối it, như vậy nếu gọi nhiều lần một biến let trong một khối it thì giá trị của nó vẫn không đổi
Còn đối với let! thì giá trị được tính toán trong khối before ngầm do đó giá trị trong các khối it giống nhau

Reply:
ĐÚng rồi á vì tính chất luôn luôn chạy vào trong 1 khối before trước mỗi test case nên nó sẽ tốn rất nhiều tài nguyên để load nhé, nên em hạn chế dùng let! nhé. Chỗ này bỏ ! có chạy không em

Reply:
@anlvq-1058 vẫn chạy đc a ạ, em fix lại r